modules/caddyhttp/caddyhttp.go

```diff
@@ -244,6 +245,40 @@ func SanitizedPathJoin(root, reqPath string) string {
 	return path
 }
 
+// CleanPath cleans path p according to path.Clean(), but only
+// merges repeated slashes if collapseSlashes is true, and always
+// preserves trailing slashes.
+func CleanPath(p string, collapseSlashes bool) string {
```

Comment:
I gave this another thought.

Typically I guess [dot segments](https://www.rfc-editor.org/rfc/rfc3986#section-5.2.4) ("." and "..") are rarely used in URLs, but I'm afraid the behavior of `CleanPath` is hard to reason about if  dot segments are used along with multiple slashes.

For example, I'm not sure if `/foo//./bar` is the result most users expect 🤔

| p | CleanPath | path.Clean |
| --- | --- | --- |
| /foo//./..//./bar | /foo//./bar | /bar |
| /foo/./.././bar | /bar | /bar |

Reply:
Interesting! I'll look into this...

Reply:
I guess we need to decide if `//..` is `/` or `//`.

Reply:
Oh, this is an important point!

Obviously, [path.Clean](https://pkg.go.dev/path#Clean) chooses `/`, but personally I think `/foo` is a better choice because [URI Empty Path Segments Matter](http://davescoolblog.blogspot.com/2011/11/uri-empty-path-segments-matter.html).

Reply:
The same author made an interesting point about path matching in [Application Content URI Rules wildcard syntax](http://davescoolblog.blogspot.com/2017/05/application-content-uri-rules-wildcard.html):

> For the path, an asterisk matches within the path segment. For example, http://example.com/a/*/c will match http://example.com/a/b/c and http://example.com/a//c but not http://example.com/a/b/b/c or http://example.com/a/c

Reply:
@RussellLuo Thanks for those resources. I think I tend to agree, if we're not merging slashes, then treat empty path components as significant. I pushed a commit just now which makes the new test case pass.